### PR TITLE
Include a list of enabled features in the register request

### PR DESCRIPTION
**Pull request by moskyb**
**This PR:** Adds a smallish list of feature enabled to the agent register request, allowing us at buildkite a bit more visibility into what features of the agent our users actually used. This telemetry is on by default, but doesn't collect any sensitive info, and can be disabled using the `--no-telemetry` flag.

### CODE REVIEW

File: api/agents.go [hunk 1]
```diff
@@ -2,18 +2,29 @@ package api
...
+}
+
+type FeatureUsage struct {
+	TracingBackend             string   `json:"tracing_backend"`
+	Experiments                []string `json:"experiments"`
+	DisconnectAfterJobEnabled  bool     `json:"disconnect_after_job_enabled"`
+	DisconnectAfterIdleEnabled bool     `json:"disconnect_after_idle_timeout_enabled"`
+	Shell                      string   `json:"shell"`
+	PluginsEnabled             bool     `json:"plugins_enabled"`
+	AcquireModeEnabled         bool     `json:"acquire_mode_enabled"`
```
Comment (pda) [inline api/agents.go:27]: Maybe this?

```suggestion
	AcquireJobEnabled         bool     `json:"acquire_job_enabled"`
```

**Review by pda — APPROVED** [review]
Excellent 👌🏼

I guess we'll want to add it to some docs / README / something too, before it lands in a release.

File: api/agents.go [hunk 2]
```diff
@@ -2,18 +2,29 @@ package api
...
+	Arch               string        `json:"arch"`
+	ScriptEvalEnabled  bool          `json:"script_eval_enabled"`
+	IgnoreInDispatches bool          `json:"ignore_in_dispatches"`
+	Priority           string        `json:"priority,omitempty"`
+	Version            string        `json:"version"`
+	Build              string        `json:"build"`
+	Tags               []string      `json:"meta_data"`
+	PID                int           `json:"pid,omitempty"`
+	MachineID          string        `json:"machine_id,omitempty"`
+	FeatureUsage       *FeatureUsage `json:"feature_usage"`
```
Comment (sj26) [inline api/agents.go:17]: Could we call this "telemetry" instead? Feels like a more general concept, and matches the arg name.

File: api/agents.go [hunk 3]
```diff
@@ -2,18 +2,29 @@ package api
...
+}
+
+type FeatureUsage struct {
+	TracingBackend             string   `json:"tracing_backend"`
```
Comment (sj26) [inline api/agents.go:21]: Should this be omitted when empty?

```suggestion
	TracingBackend             string   `json:"tracing_backend,omitempty"`
```

File: api/agents.go [hunk 4]
```diff
@@ -2,18 +2,29 @@ package api
...
+	Experiments                []string `json:"experiments"`
```
Comment (sj26) [inline api/agents.go:22]: Nice

File: api/agents.go (hunk 3 above)
Comment (moskyb) [reply api/agents.go:21]: ah yes! it should, it's been so long since i serialised a struct, i completely forgot.

File: api/agents.go (hunk 2 above)
Comment (moskyb) [reply api/agents.go:17]: eh... i think telemetry covers a larger concept that what we have here - ie, if we were sending back job timings to buildkite (for buildkite to see and not the client), that would also be another type of telemetry, that i think we would want to disable that using the same flag that disables this. Does that make sense? I feel like i haven't explained it that well ahah

Comment (sj26) [reply api/agents.go:17]: Yeah, true. I wish we didn't already use the term "meta_data" 😬 

Maybe feature_usage works. It just feels odd in the backend.

Comment (moskyb) [reply api/agents.go:17]: maybe `agent_feature_usage` is a little bit less odd?

File: api/agents.go (hunk 3 above)
Comment (moskyb) [reply api/agents.go:21]: actually, thinking on this, i think it makes sense to leave that as `""` if there's no tracing backend configured - i think it's marginally clearer in that case that tracing is disabled?

the other option would be to have two things in here, `tracing_enabled bool` and `tracing_backend bool omitempty`

Comment (pda) [reply api/agents.go:21]: Or, `open_tracing_enabled` and `open_telemetry_enabled` bools.
I think that's my favourite.

File: api/agents.go (hunk 2 above)
Comment (sj26) [reply api/agents.go:17]: Maybe the other direction, what about just `features`?

Comment (sj26) [reply api/agents.go:17]: I'm overthinking it.

I've adjusted the backend to accept `feature_usage` as a json blob.

**Review by sj26 — APPROVED** [review]
Nice, and thanks!

File: api/agents.go [hunk 5]
```diff
@@ -2,18 +2,29 @@ package api
...
+	Arch               string        `json:"arch"`
+	ScriptEvalEnabled  bool          `json:"script_eval_enabled"`
+	IgnoreInDispatches bool          `json:"ignore_in_dispatches"`
+	Priority           string        `json:"priority,omitempty"`
+	Version            string        `json:"version"`
+	Build              string        `json:"build"`
+	Tags               []string      `json:"meta_data"`
+	PID                int           `json:"pid,omitempty"`
+	MachineID          string        `json:"machine_id,omitempty"`
+	FeatureUsage       *FeatureUsage `json:"feature_usage,omitempty"`
```
Comment (sj26) [inline api/agents.go:17]: I'd like to call this parameter `features`, and make it an array of simple strings.

File: clicommand/agent_start.go [hunk 6]
```diff
@@ -94,6 +94,7 @@ type AgentStartConfig struct {
 	NoPlugins                   bool     `cli:"no-plugins"`
 	NoPluginValidation          bool     `cli:"no-plugin-validation"`
 	NoPTY                       bool     `cli:"no-pty"`
+	NoTelemetry                 bool     `cli:"no-telemetry"`
```
Comment (sj26) [inline clicommand/agent_start.go:97]: I thin "no telemetry" is too broad. Folks might expect us not to send operating system or process information under this name.

I'd suggest instead we use `--no-features-reporting` to be more specific.

File: clicommand/agent_start.go [hunk 7]
```diff
@@ -414,6 +457,11 @@ var AgentStartCommand = cli.Command{
 			Usage:  "Send metrics to DogStatsD for Datadog",
 			EnvVar: "BUILDKITE_METRICS_DATADOG",
 		},
+		cli.BoolFlag{
+			Name:   "no-telemetry",
+			Usage:  "Disables sending usage statistics back to the buildkite mothership. We use the information gathered from telemetry to figure out what features are getting used.",
+			EnvVar: "BUILDKITE_AGENT_NO_TELEMETRY",
+		},
```
Comment (sj26) [inline clicommand/agent_start.go:464]: This one needs to be updated I think?

**Review by sj26 — APPROVED** [review]
Looks good! I noticed the bool flag didn't catch up on the rename, but otherwise this looks great to go

File: clicommand/agent_start.go [hunk 8]
```diff
@@ -130,6 +131,48 @@ type AgentStartConfig struct {
...
+	if asc.GitMirrorsPath != "" {
+		features = append(features, "git-mirrors")
+	}
+
+	if asc.AcquireJob != "" {
+		features = append(features, "acquire-job")
+	}
+
+	if asc.TracingBackend == tracetools.BackendDatadog {
+		features = append(features, "tracing-datadog")
```
Comment (sj26) [inline clicommand/agent_start.go:146]:
```suggestion
		features = append(features, "datadog-tracing")
```

File: clicommand/agent_start.go [hunk 9]
```diff
@@ -130,6 +131,48 @@ type AgentStartConfig struct {
...
+	}
+
+	if asc.TracingBackend == tracetools.BackendOpenTelemetry {
+		features = append(features, "tracing-opentelemetry")
```
Comment (sj26) [inline clicommand/agent_start.go:150]:
```suggestion
		features = append(features, "opentelemetry-tracing")
```

**Comment by sj26** [comment]
I tried this out end-to-end against a development version of Buildkite and observed it recording feature usage 🙌

**Review by pda — APPROVED** [review]
Looks good to me, I've tested it locally.
Ideally some tests would be nice. Not sure if we have an existing place to put them for this layer of the agent, though.

**Pull request merged**